text/0000-module-unification.md

```diff
@@ -0,0 +1,837 @@
+
+* Resolution rules that are declared only in JavaScript are difficult to
+  analyze and optimize.
+
+* Module resolution is inefficient due to the number of potential places to
+  lookup a particular module by name.
+
+Recognizing these drawbacks, the Core Team compiled a set of
+[design constraints](https://github.com/emberjs/core-notes/blob/master/ember.js/2016-01/january-22.md#summary-of-pods-design-constraints)
+for a rethink of Ember's approach to modules:
+
+1. Reasonable branching factor. Users should see a reasonable number of items at any given level in their hierarchy. Flattening out too much results in an unreasonably large number of items.
+2. No slashes in component names. The existing system allows this, but we don't want to support invocation of nested components in Glimmer Components.
+3. Addons need to participate in the naming scheme, most likely with namespace prefix and double-colon separator.
+4. Subtrees should be relocatable. If you move a directory to a new place in the tree, its internal structure should all still work.
+5. There should be no cognitive overhead when adding a new thing. The right way should be obvious and not impose unnecessary decisions.
+6. We need clean separation between the namespace of the user's own components, helpers, routes, etc and the framework's own type names ("component", "helper", etc) so that we can disambiguate them and add future ones.
+7. Ideally we will have a place to put tests and styles alongside corresponding components.
+8. Local relative lookup for components and helpers needs to work.
+9. Avoid the "titlebar problem", in which many files are all named "component.js" and you can't tell them apart in your editor.
+10. The resolver should be configured via declarative rules, not imperative
+    JavaScript. In addition to enforcing consistency, this allows addons to
+    augment the system with their own types in a predictable way.
+11. Module structures must be statically analyzable at build time to enable
+    efficiency optimizations.
+12. Module classifications must be extensible and allow for customizations by
+    apps, engines, and addons.
+
+> Note: Constraints > 9 were added based on discussions subsequent to the
+initial meeting.
+
+This proposal attempts to address these constraints with a single consistent
+approach to modules that will make Ember easier to use and learn _and_ improve
+the efficiency of Ember's resolver at run-time.
+
+# Detailed Design
+
+This proposal introduces a new top-level directory, `src`, and establishes
+conventions for organizing modules within it. Also proposed is a refactor of the
+Ember resolver to enable efficient and flexible resolutions based upon the new
+module conventions.
+
+The `src` directory will be used to contain the core ES modules within an Ember
+CLI project, whether that project contains an application, addon, or engine. To
+maintain backward compatibility, the `src` directory will be allowed to
```

Comment:
@bcardarella One of the goals of the new module layout is that it can be migrated to incrementally. Given your experiences in the past feeling stuck straddling two versions, I'm sure this is a goal you share as well.